README.md

```diff
@@ -0,0 +1,306 @@
+---
+title: Nested dependencies in frontend
+published: false
+description: What are nested dependencies, why they exist, how they can harm the frontend and what you can do to solve it.
+tags: javascript, import-maps, node, rollup
+---
+
+So you got this awesome idea and you want to now actually do it.
+I am pretty sure you do not want to start from scratch so let's use existing open source packages.
+
+If you want to play along - all the code is on [github](https://github.com/daKmoR/nested-dependecies-in-frontend).
+
+For our example case, we wanna use lit-element and lit-html.
+
+```bash
+mkdir nested-dependecies-in-frontend
+cd nested-dependecies-in-frontend
+npm install lit-element lit-html@1.0.0 --save-exact
+```
+
+> we are using pinned versions on purpose here
+
+and then we just load both packages in our `main.js`.
+```js
+import { LitElement } from "lit-element";
+import { html } from "lit-html";
+
+console.log(LitElement);
+console.log(html);
+```
+
+In order to find out how big our app will be, we would like to create a rollup bundle.
+
+```bash
+npm install -D rollup
+```
+and create a `rollup.config.js`
+```js
+export default {
+  input: "main.js",
+  output: {
+    file: "bundle.js",
+    format: "iife"
+  },
+};
+```
+and add a `"build": "rollup -c rollup.config.js && du -h bundle.js"` to our package.json so we build the file and output it's file size.
+Let's run it :)
+
+oh it doesn't work :sob:
+
+```
+(!) Unresolved dependencies
+https://rollupjs.org/guide/en#warning-treating-module-as-external-dependency
+lit-element (imported by main.js)
+lit-html (imported by main.js)
+```
+
+ok, I heard that before... we need to add some plugins to get it to understand the way node resolution works.
+
+```bash
+npm i -D rollup-plugin-node-resolve
+```
+
+modify our rollup to add
+```js
+import resolve from "rollup-plugin-node-resolve";
+
+export default {
+  input: "main.js",
+  output: {
+    file: "bundle.js",
+    format: "iife"
+  },
+  plugins: [resolve()]
+};
+```
+
+```bash
+$ npm run build
+# ...
+created bundle.js in 414ms
+96K     bundle.js
+```
+
+So that seems to work fine :muscle:
+
+### What happens if someone prefers yarn?
+
+Doing a yarn install and then a build should result in the same output right?
+
+```bash
+$ yarn install
+$ yarn build
+# ...
+created bundle.js in 583ms
+124K    bundle.js
+```
+
+#### wow that is unexpected - 124K vs 96K?
+
+It seems yarn has some extra files... maybe a package is double?
+
+```bash
+$ yarn list --pattern lit-*
+├─ lit-element@2.2.0
+│  └─ lit-html@1.1.0
+└─ lit-html@1.0.0
+```
+
+jup version `1.0.0` and `1.1.0` is available for `lit-html`.
+The reason is most likely that we have a pinned `1.0.0` version in our root dependency.
+
+npm seems to dedupe it fine...
```

Comment:
```suggestion
While `npm` seems to dedupe it fine, I don't feel safe using `npm` because if the dependency tree becomes bigger npm also likes to install nested dependencies.
```